tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/Multigrid/Helpers.hpp

```diff
@@ -0,0 +1,252 @@
+        db::get<::domain::Tags::InitialExtents<1>>(box);
+    const auto& domain = db::get<::domain::Tags::Domain<1>>(box);
+    auto mesh = ::domain::Initialization::create_initial_mesh(
+        initial_extents, element_id, Spectral::Quadrature::GaussLobatto);
+    const auto logical_coords = logical_coordinates(mesh);
+    const auto& block = domain.blocks()[element_id.block_id()];
+    const ElementMap<1, Frame::Inertial> element_map{
+        element_id, block.stationary_map().get_clone()};
+    auto inertial_coords = element_map(logical_coords);
+    // Initialize data
+    const size_t element_index = helpers_distributed::get_index(element_id);
+    const size_t mg_lev = element_id.grid_index();
+    const auto& source =
+        mg_lev == 0 ? typename sources_tag::type(gsl::at(
+                          get<helpers_distributed::Source>(box), element_index))
+                    : typename sources_tag::type{};
+    const size_t num_points = mesh.number_of_grid_points();
+    auto initial_fields = typename fields_tag::type{num_points, 0.};
+    Initialization::mutate_assign<simple_tags>(
+        make_not_null(&box), std::move(mesh), std::move(inertial_coords),
+        std::move(initial_fields), std::move(source));
+    return {std::move(box)};
+  }
+};
+
+// The next two actions are responsible for applying the linear operator to the
+// operand data. They apply a chunk of the full matrix on each element and
+// perform a reduction to obtain the full matrix-vector product.
+
+template <typename OperandTag>
+struct CollectOperatorAction;
+
+template <typename OperandTag>
+struct ComputeOperatorAction {
+  using const_global_cache_tags = tmpl::list<LinearOperator>;
+  using local_operator_applied_to_operand_tag =
+      db::add_tag_prefix<::LinearSolver::Tags::OperatorAppliedTo, OperandTag>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ActionList, typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&, Parallel::AlgorithmExecution>
+  apply(db::DataBox<DbTagsList>& box,
+        const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+        const Parallel::GlobalCache<Metavariables>& /*cache*/,
+        const ElementId<1>& element_id, const ActionList /*meta*/,
+        const ParallelComponent* const /*meta*/) noexcept {
+    const size_t mg_lev = element_id.grid_index();
+    const size_t element_index = helpers_distributed::get_index(element_id);
+    const auto& operator_matrices = get<LinearOperator>(box)[mg_lev];
+    const size_t number_of_elements = operator_matrices.size();
+    const auto& linear_operator = gsl::at(operator_matrices, element_index);
+    const size_t number_of_grid_points = linear_operator.columns();
+    const auto& operand = get<OperandTag>(box);
+
+    typename OperandTag::type operator_applied_to_operand{
+        number_of_grid_points * number_of_elements};
+    dgemv_('N', linear_operator.rows(), linear_operator.columns(), 1,
```

Comment:
Can you use `ApplyMatrices` for this?

Reply:
I had to add a commit that make `Matrix` a type alias (instead of a derived class) for this to work with both option-construction and `apply_matrices`. I pushed the commit here and also to #3374.

Reply:
Actually, this leads down a whole rabbit hole related to Blaze-interoperability. If you don't mind I'll leave the explicit blas call here fore now and deal with the Blaze issues separately.

Reply:
Okay, you can leave it as-is, then, just add a code comment explaining the reason you can't use `apply_matrices`.